libpod/lock/shm/shm_lock.c

```diff
@@ -0,0 +1,383 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <semaphore.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "shm_lock.h"
+
+// Compute the size of the SHM struct
+size_t compute_shm_size(uint32_t num_bitmaps) {
+  return sizeof(shm_struct_t) + (num_bitmaps * sizeof(lock_group_t));
+}
+
+// Set up an SHM segment holding locks for libpod.
+// num_locks must be a multiple of BITMAP_SIZE (32 by default).
+// Returns a valid pointer on success or NULL on error.
+// If an error occurs, it will be written to the int pointed to by error_code.
+shm_struct_t *setup_lock_shm(uint32_t num_locks, int *error_code) {
+  int shm_fd, i, j, ret_code;
+  uint32_t num_bitmaps;
+  size_t shm_size;
+  shm_struct_t *shm;
+
+  // If error_code doesn't point to anything, we can't reasonably return errors
+  // So fail immediately
+  if (error_code == NULL) {
+    return NULL;
+  }
+
+  // We need a nonzero number of locks
+  if (num_locks == 0) {
+    *error_code = EINVAL;
+    return NULL;
+  }
+
+  // Calculate the number of bitmaps required
+  if (num_locks % BITMAP_SIZE != 0) {
+    // Number of locks not a multiple of BITMAP_SIZE
+    *error_code = EINVAL;
+    return NULL;
+  }
+  num_bitmaps = num_locks / BITMAP_SIZE;
+
+  // Calculate size of the shm segment
+  shm_size = compute_shm_size(num_bitmaps);
+
+  // Create a new SHM segment for us
+  shm_fd = shm_open(SHM_NAME, O_RDWR | O_CREAT | O_EXCL, 0600);
+  if (shm_fd < 0) {
+    *error_code = errno;
+    return NULL;
+  }
+
+  // Increase its size to what we need
+  ret_code = ftruncate(shm_fd, shm_size);
+  if (ret_code < 0) {
+    *error_code = errno;
+    goto CLEANUP_UNLINK;
+  }
+
+  // Map the shared memory in
+  shm = mmap(NULL, shm_size, PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
+  if (shm == MAP_FAILED) {
+    *error_code = errno;
+    goto CLEANUP_UNLINK;
+  }
+
+  // We have successfully mapped the memory, now initialize the region
+  shm->magic = MAGIC;
+  shm->num_locks = num_locks;
+  shm->num_bitmaps = num_bitmaps;
+
+  // Initialize the semaphore that protects the bitmaps.
+  // Initialize to value 1, as we're a mutex, and set pshared as this will be
+  // shared between processes in an SHM.
+  ret_code = sem_init(&(shm->segment_lock), true, 1);
```

Comment:
Not really: If we needed to worry about that race, we also need to worry about the other caller accessing the semaphore before it is initialized at all, and trying to lock garbage. Sure, it would make the race window a bit smaller, but it would still exist, and be more difficult to spot to a casual reader; I’m not sure that’s a net win.

(Also, this lock is used only for exclusivity when allocating other locks, it is not required for the other callers to just use the other locks by pre-existing IDs.)